### PR TITLE
fix paralel image copy

### PR DESCRIPTION
**Pull request by tiborpopovics**

### CODE REVIEW

**Comment by pdarvasi** [comment]
I've retriggered the IT

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureStorageAccountService.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import com.microsoft.azure.management.storage.StorageAccount;
+import com.sequenceiq.cloudbreak.cloud.azure.client.AzureClient;
+import com.sequenceiq.cloudbreak.cloud.azure.task.AzurePollTaskFactory;
+import com.sequenceiq.cloudbreak.cloud.azure.task.storageaccount.StorageAccountCheckerContext;
+import com.sequenceiq.cloudbreak.cloud.context.AuthenticatedContext;
+import com.sequenceiq.cloudbreak.cloud.model.CloudStack;
+import com.sequenceiq.cloudbreak.cloud.scheduler.SyncPollingScheduler;
+import com.sequenceiq.cloudbreak.cloud.task.PollTask;
+
+@Component
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureStorageAccountService.java:22]: pls use @Service for services

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureStorageAccountService.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    public void createStorageAccount(AuthenticatedContext ac, AzureClient client, String resourceGroup, String storageName, String region, CloudStack stack) {
+        StorageAccount storageAccount = client.getStorageAccountByGroup(resourceGroup, storageName);
+        if (storageAccount == null) {
+            try {
+                LOGGER.info("Creating storage account: {}", storageName);
+                armStorage.createStorage(client, storageName, LOCALLY_REDUNDANT, resourceGroup, region, isEncryptionNeeded(stack), stack.getTags());
+                pollStorageAccountCreation(ac, new StorageAccountCheckerContext(client, resourceGroup, storageName));
+                LOGGER.info("Creating container: {} in storage storage account: {}", IMAGES_CONTAINER, storageName);
+                client.createContainerInStorage(resourceGroup, storageName, IMAGES_CONTAINER);
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureStorageAccountService.java:48]: pls. use a separate method for this or rename this one to indicate that it is doing this, too

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureStorageAccountService.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            } catch (CloudException e) {
+                LOGGER.warn("Error during storage account creation: {}.", storageName, e);
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureStorageAccountService.java:50]: This causes the deployment to fail, don't we want to rethrow or retry?

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureStorageAccountService.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            LOGGER.info("Storage account {} already exist in resource group {}", storageName, resourceGroup);
+        }
+    }
+
+    private void pollStorageAccountCreation(AuthenticatedContext ac, StorageAccountCheckerContext checkerContext) {
+        PollTask<Boolean> storageAccountCreationStatusCheckerTask = azurePollTaskFactory.storageAccountCheckerTask(ac, checkerContext);
+        try {
+            LOGGER.info("Start polling storage account creation: {}", checkerContext.getStorageAccountName());
+            syncPollingScheduler.schedule(storageAccountCreationStatusCheckerTask, STORAGE_ACCOUNT_CREATION_CHECKING_INTERVAL,
+                    STORAGE_ACCOUNT_CREATION_CHECKING_MAX_ATTEMPT, 1);
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureStorageAccountService.java:62]: Do we want to tolerate only 1 failure?

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import com.sequenceiq.cloudbreak.cloud.azure.view.AzureCredentialView;
+import com.sequenceiq.cloudbreak.cloud.context.AuthenticatedContext;
+import com.sequenceiq.cloudbreak.cloud.context.CloudContext;
+import com.sequenceiq.cloudbreak.cloud.exception.CloudConnectorException;
+import com.sequenceiq.cloudbreak.cloud.model.CloudStack;
+import com.sequenceiq.cloudbreak.cloud.model.Image;
+import com.sequenceiq.common.api.type.ImageStatus;
+import com.sequenceiq.common.api.type.ImageStatusResult;
+
+@Component
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java:30]: pls use @service for services

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        String imageResourceGroupName = azureResourceGroupMetadataProvider.getImageResourceGroupName(ac.getCloudContext(), stack);
+        AzureClient client = ac.getParameter(AzureClient.class);
+
+        AzureCredentialView acv = new AzureCredentialView(ac.getCloudCredential());
+        String imageStorageName = armStorage.getImageStorageName(acv, ac.getCloudContext(), stack);
+        try {
+            CopyState copyState = client.getCopyStatus(imageResourceGroupName, imageStorageName, IMAGES_CONTAINER, image.getImageName());
+            if (CopyStatus.SUCCESS.equals(copyState.getStatus())) {
+                if (StringUtils.isEmpty(armStorage.getCustomImageId(client, ac, stack))) {
+                    return new ImageStatusResult(ImageStatus.CREATE_FAILED, ImageStatusResult.COMPLETED);
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java:54]: pls. log this

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                }
+                return new ImageStatusResult(ImageStatus.CREATE_FINISHED, ImageStatusResult.COMPLETED);
+            } else if (CopyStatus.ABORTED.equals(copyState.getStatus()) || CopyStatus.INVALID.equals(copyState.getStatus())) {
+                return new ImageStatusResult(ImageStatus.CREATE_FAILED, 0);
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java:58]: pls log this

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    public void copyVhdImageIfNecessary(AuthenticatedContext ac, CloudStack stack, Image image, String region, AzureClient client) {
+        CloudContext cloudContext = ac.getCloudContext();
+        String resourceGroupName = azureResourceGroupMetadataProvider.getResourceGroupName(cloudContext, stack);
+        String imageStorageName = armStorage.getImageStorageName(new AzureCredentialView(ac.getCloudCredential()), cloudContext, stack);
+        String imageResourceGroupName = azureResourceGroupMetadataProvider.getImageResourceGroupName(cloudContext, stack);
+
+        createResourceGroupIfNotExists(client, resourceGroupName, region, stack);
+        createResourceGroupIfNotExists(client, imageResourceGroupName, region, stack);
+        azureStorageAccountService.createStorageAccount(ac, client, imageResourceGroupName, imageStorageName, region, stack);
+        if (!storageContainsImage(client, imageResourceGroupName, imageStorageName, image.getImageName())) {
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java:81]: pls log the "else" cause of this if

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            } catch (CloudConnectorException e) {
+                LOGGER.warn("Something happened during start image copy.", e);
+            }
+        }
+    }
+
+    private boolean storageContainsImage(AzureClient client, String resourceGroupName, String storageName, String image) {
+        List<ListBlobItem> listBlobItems = client.listBlobInStorage(resourceGroupName, storageName, IMAGES_CONTAINER);
+        for (ListBlobItem listBlobItem : listBlobItems) {
+            if (getNameFromConnectionString(listBlobItem.getUri().getPath()).equals(image.split("/")[image.split("/").length - 1])) {
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/AzureImageSetupService.java:94]: you can use getNameFromConnectionString for the image, too

**Comment by keyki** [comment]
Can you please fix the commit message to include the jira ticket?

**Comment by pdarvasi** [comment]
> Can you please fix the commit message to include the jira ticket?

I am merging to make progress as this does not affect the functionality.

**Pull request merged**